src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs

```diff
@@ -1595,25 +1604,40 @@ internal static void CheckActionPreference(FunctionContext funcContext, Exceptio
             // set $? to false indicating an error
             context.QuestionMarkVariableValue = false;
 
-            bool anyTrapHandlers = funcContext._traps.Any() && funcContext._traps.Last().Item2 != null;
+            ActionPreference preference = GetErrorActionPreference(context);
 
-            if (!anyTrapHandlers && !ExceptionHandlingOps.NeedToQueryForActionPreference(rte, context))
+            if (IsExceptionNew(rte, context) &&
+                preference == ActionPreference.Break)
             {
-                throw rte;
+                context.Debugger?.Break(rte);
+
+                // set the value of $? here in case it was reset in the debugger.
+                context.QuestionMarkVariableValue = false;
             }
 
-            ActionPreference preference;
+            // Item2 in the trap tuples is the action (script) for the trap.
+            // A null action script is only used to indicate when exceptions
+            // should be thrown up to a higher level, and doesn't count as an
+            // actual trap handler in the function context.
+            bool anyTrapHandlers = funcContext._traps.Count > 0 && funcContext._traps[funcContext._traps.Count - 1].Item2 != null;
+
             if (anyTrapHandlers)
             {
+                // update the action preference according to how the exception is
+                // handled in the trap statement(s).
                 preference = ProcessTraps(funcContext, rte);
+
+                // set the value of $? here in case it was reset in trap handling.
+                context.QuestionMarkVariableValue = false;
             }
-            else
+            else if (ExceptionCannotBeStoppedContinuedOrIgnored(rte, context))
             {
-                preference = ExceptionHandlingOps.QueryForAction(rte, rte.Message, context);
+                throw rte;
+            }
+            else if (preference == ActionPreference.Inquire && !rte.SuppressPromptInInterpreter)
+            {
+                preference = InquireForActionPreference(rte.Message, context);
```

Comment:
It looks like `QuestionMarkVariableValue` is not being set here.  Should it be?

Reply:
Yes, it should be. Good catch.

Reply:
FYI, this is fixed @PaulHigin.